### PR TITLE
KAFKA-9020: Streams sub-topologies should be sorted by sink -> source relationship

### PR DESCRIPTION
**Pull request by ableegoldman**
Subtopologies are currently ordered alphabetically by source node, which prior to KIP-307 happened to always result in the "correct" (ie topological) order. Now that users may name their nodes anything they want, we must explicitly order them so that upstream node groups/subtopologies come first and the downstream ones come after.

### CODE REVIEW

**Comment by ableegoldman** [comment]
@guozhangwang @bbejeck @vvcephei

**Comment by guozhangwang** [comment]
ping @bbejeck @vvcephei could you guys take a look?

**Review by bbejeck — APPROVED** [review]
Thanks for this fix @ableegoldman! LGTM

**Comment by bbejeck** [comment]
Java 11/2.12 failed with `Task :connect:mirror:integrationTest FAILED`

Java 11/2.13 failed with `kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup`

Java 8 failed with `FATAL: Unable to delete script file /tmp/jenkins6761014166594532618.sh
06:53:19 java.io.EOFException`

retest this please

**Comment by ableegoldman** [comment]
Java 8 and Java 11.12 passed
Java 11 says this failed but when you click to see the stacktrace it says it passed, sooo...?
`KTableKTableForeignKeyJoinIntegrationTest.doLeftJoinFilterOutRapidlyChangingForeignKeyValues`

**Comment by bbejeck** [comment]
ping either of @guozhangwang @vvcephei for a second review

**Comment by bbejeck** [comment]
Java 11/2.13 failed  - test results already cleaned up.
Java 11/2.12 and Java 8 passed

retest this please

**Review by guozhangwang — APPROVED** [review]
LGTM.

**Comment by bbejeck** [comment]
on this PR builder run Java 11/2.12 and Java 11/2.13 passed
on previous PR builder run Java 8 and Java 11/2.12 passed, so between the two all passed.
Merging this now.

**Comment by bbejeck** [comment]
Merged #7495 into trunk.

**Comment by bbejeck** [comment]
cherry-picked to 2.4

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java [hunk 1]
```diff
@@ -749,27 +748,18 @@ private void connectStateStoreNameToSourceTopicsOrPattern(final String stateStor
...
 
         int nodeGroupId = 0;
 
-        // Go through source nodes first. This makes the group id assignment easy to predict in tests
-        final Set<String> allSourceNodes = new HashSet<>(nodeToSourceTopics.keySet());
-        allSourceNodes.addAll(nodeToSourcePatterns.keySet());
-
-        for (final String nodeName : Utils.sorted(allSourceNodes)) {
+        // Traverse in topological order
+        for (final String nodeName : nodeFactories.keySet()) {
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:759]: q: how could we guarantee a topological traversal on a map without order?

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:759]: `nodeFactories` is in topological order, for example in `InternalStreamsBuilder#buildAndOptimizeTopology` we only add a node once all its parents have been written to the topology already

**Pull request merged**